### PR TITLE
Reservation jobs might run on non reservation nodes

### PR DESCRIPTION
**Pull request by arungrover**
#### Describe Bug or Feature
Sometimes when a degraded reservation starts running and when it has no nodes associated to it, the jobs in these reservations might run on any available nodes.
This issue can happen only when all the reservation's nodes are taken away from reservation before it could start. Maintenance reservations can cause this issue when it completely takes over the nodes from and existing confirmed reservation, leaving the confirmed reservation with no nodes at all.

#### Describe Your Change

The scheduler already detects and ignores any running reservation that does not have any reserved nodes in them. But it does not mark the jobs in the reservation queues as can-not-run. This makes the scheduler consider those queued jobs as any other normal queued job and makes the job run on nodes that are not part of the reservation.
After this change, when the scheduler ignores the reservations, it will also mark all the jobs in the associated queue as can-never-run.

#### Link to Design Doc
NA

#### Attach Test and Valgrind Logs/Output
[test.txt](https://github.com/openpbs/openpbs/files/4848949/test.txt)

### CODE REVIEW

File: test/tests/functional/pbs_maintenance_reservations.py [hunk 1]
```diff
@@ -639,3 +639,58 @@ def test_maintenance_progressive_degrade_reservation(self):
         time.sleep(30)
 
         self.server.expect(JOB, {'job_state': 'Q'}, id=jid1)
+
+    def test_maintenance_degrade_reservation_jobs_dont_run(self):
```
Comment (vstumpf) [inline test/tests/functional/pbs_maintenance_reservations.py:643]: use the @requirements decorator?

File: test/tests/functional/pbs_maintenance_reservations.py [hunk 2]
```diff
@@ -639,3 +639,58 @@ def test_maintenance_progressive_degrade_reservation(self):
...
+        resv_state = {'reserve_state': (MATCH_RE, 'RESV_RUNNING|5')}
+        self.logger.info('Sleeping until reservation starts')
+        offset = start - int(time.time())
+        self.server.expect(RESV, resv_state, id=rid,
+                           offset=offset)
+        self.server.expect(RESV, resv_state, id=rid2,
+                           offset=offset)
+
+        self.logger.info('wait 5 seconds to make sure scheduler looks at job')
+        time.sleep(5)
```
Comment (vstumpf) [inline test/tests/functional/pbs_maintenance_reservations.py:693]: maybe use offset instead of a raw sleep?

File: src/scheduler/resv_info.c [hunk 3]
```diff
@@ -132,14 +133,15 @@ stat_resvs(int pbs_sd)
...
+ * @param[in] resvs  - batch status of the stat'ed reservations
+ * @param[in] pbs_sd - connection to the pbs server
  *
- *	returns an array of reservations
+ * @return    An array of reservations
  *
  */
 resource_resv **
-query_reservations(server_info *sinfo, struct batch_status *resvs)
+query_reservations(server_info *sinfo, struct batch_status *resvs, int pbs_sd)
```
Comment (bhroam) [inline src/scheduler/resv_info.c:144]: Please swap the order of arguments.  If we pass in a pbs_sd it comes first (unless policy is there, then it comes second).

File: src/scheduler/resv_info.c [hunk 4]
```diff
@@ -239,6 +241,16 @@ query_reservations(server_info *sinfo, struct batch_status *resvs)
 
 		if (ignore_resv == 1) {
 			sinfo->num_resvs--;
+			/* mark all the jobs of the associated queue as can never run */
+			if (resresv->resv->queuename != NULL) {
+				queue_info *qinfo = find_queue_info(sinfo->queues, resresv->resv->queuename);
+				if (qinfo != NULL) {
+				    clear_schd_error(err);
+				    set_schd_error_arg(err, SPECMSG, "Reservation is in invalid state");
```
Comment (bhroam) [inline src/scheduler/resv_info.c:249]: Either make this 'Reservation is in an invalid state' or 'Reservation is invalid'

File: src/scheduler/resv_info.h [hunk 5]
```diff
@@ -55,7 +55,7 @@ struct batch_status *stat_resvs(int pbs_sd);
 /*
  *	query_reservations - query the reservations from the server
  */
-resource_resv **query_reservations(server_info *sinfo, struct batch_status *resvs);
+resource_resv **query_reservations(server_info *sinfo, struct batch_status *resvsi, int pbs_sd);
```
[the remark this answers is not in the document]
Comment (arungrover) [reply src/scheduler/resv_info.h:58]: oops

File: test/tests/functional/pbs_maintenance_reservations.py (hunk 2 above)
Comment (arungrover) [reply test/tests/functional/pbs_maintenance_reservations.py:693]: Based on Bhroam's comments, I'll just trigger cycle and remove sleep.

**Comment by arungrover** [comment]
[test.txt](https://github.com/openpbs/openpbs/files/4886690/test.txt)

**Comment by arungrover** [comment]
Thanks @bhroam and @vstumpf for reviewing the PR.
It's good to go.

**Pull request merged**